File: stacknqueue/stack-using-array.py
class Stack:
    def __init__(self, capacity):
        self.capacity = capacity
        self.stack = []

    def is_empty(self):
        return len(self.stack) == 0

    def is_full(self):
        return len(self.stack) == self.capacity

    def push(self, item):
        if self.is_full():
            print("Stack Overflow: Cannot push item, stack is full.")
        else:
            self.stack.append(item)
            print(f"Pushed {item} to the stack.")

    def pop(self):
        if self.is_empty():
            print("Stack Underflow: Cannot pop item, stack is empty.")
            return None
        else:
            return self.stack.pop()

    def peek(self):
        if self.is_empty():
            print("Stack is empty.")
            return None
        else:
            return self.stack[-1]

    def size(self):
        return len(self.stack)

    def display(self):
        if self.is_empty():
            print("Stack is empty.")
        else:
            print("Stack contents:", end=" ")
            for item in reversed(self.stack):
                print(item, end=" ")
            print()

# Function to take input from user and perform stack operations
def perform_stack_operations():
    capacity = int(input("Enter the capacity of the stack: "))
    stack = Stack(capacity)

    while True:
        print("\nChoose an operation:")
        print("1. Push")
        print("2. Pop")
        print("3. Peek")
        print("4. Check if stack is empty")
        print("5. Check if stack is full")
        print("6. Get size of stack")
        print("7. Display stack")
        print("8. Exit")

        choice = input("Enter your choice: ")

        if choice == '1':
            item = input("Enter the item to push: ")
            stack.push(item)
        elif choice == '2':
            popped_item = stack.pop()
            if popped_item is not None:
                print("Popped", popped_item)
        elif choice == '3':
            top_item = stack.peek()
            if top_item is not None:
                print("Top element of the stack:", top_item)
        elif choice == '4':
            print("Is stack empty?", stack.is_empty())
        elif choice == '5':
            print("Is stack full?", stack.is_full())
        elif choice == '6':
            print("Size of the stack:", stack.size())
        elif choice == '7':
            stack.display()
        elif choice == '8':
            print("Exiting...")
            break
        else:
            print("Invalid choice. Please choose a valid option.")

# Example usage
if __name__ == "__main__":
    perform_stack_operations()

File: stacknqueue/stack-using-array.cpp
#include <iostream>
using namespace std;

#define MAX_SIZE 100

struct Stack {
    int capacity;
    int stack[MAX_SIZE];
    int top;

    Stack(int capacity) {
        this->capacity = capacity;
        top = -1;
    }

    bool is_empty() {
        return top == -1;
    }

    bool is_full() {
        return top == capacity - 1;
    }

    void push(int item) {
        if (is_full()) {
            cout << "Stack Overflow: Cannot push item, stack is full." << endl;
        } else {
            top++;
            stack[top] = item;
            cout << "Pushed " << item << " to the stack." << endl;
        }
    }

    int pop() {
        if (is_empty()) {
            cout << "Stack Underflow: Cannot pop item, stack is empty." << endl;
            return -1;
        } else {
            int popped_item = stack[top];
            top--;
            return popped_item;
        }
    }

    int peek() {
        if (is_empty()) {
            cout << "Stack is empty." << endl;
            return -1;
        } else {
            return stack[top];
        }
    }

    int size() {
        return top + 1;
    }

    void display() {
        if (is_empty()) {
            cout << "Stack is empty." << endl;
        } else {
            cout << "Stack contents: ";
            for (int i = top; i >= 0; i--) {
                cout << stack[i] << " ";
            }
            cout << endl;
        }
    }
};

// Function to perform stack operations
void perform_stack_operations() {
    int capacity;
    cout << "Enter the capacity of the stack: ";
    cin >> capacity;
    Stack stack(capacity);

    while (true) {
        cout << "\nChoose an operation:" << endl;
        cout << "1. Push" << endl;
        cout << "2. Pop" << endl;
        cout << "3. Peek" << endl;
        cout << "4. Check if stack is empty" << endl;
        cout << "5. Check if stack is full" << endl;
        cout << "6. Get size of stack" << endl;
        cout << "7. Display stack" << endl;
        cout << "8. Exit" << endl;

        int choice;
        cout << "Enter your choice: ";
        cin >> choice;

        switch (choice) {
            case 1: {
                int item;
                cout << "Enter the item to push: ";
                cin >> item;
                stack.push(item);
                break;
            }
            case 2: {
                int popped_item = stack.pop();
                if (popped_item != -1) {
                    cout << "Popped " << popped_item << endl;
                }
                break;
            }
            case 3: {
                int top_item = stack.peek();
                if (top_item != -1) {
                    cout << "Top element of the stack: " << top_item << endl;
                }
                break;
            }
            case 4:
                cout << "Is stack empty? " << (stack.is_empty() ? "Yes" : "No") << endl;
                break;
            case 5:
                cout << "Is stack full? " << (stack.is_full() ? "Yes" : "No") << endl;
                break;
            case 6:
                cout << "Size of the stack: " << stack.size() << endl;
                break;
            case 7:
                stack.display();
                break;
            case 8:
                cout << "Exiting..." << endl;
                return;
            default:
                cout << "Invalid choice. Please choose a valid option." << endl;
        }
    }
}

// Driver code
int main() {
    perform_stack_operations();
    return 0;
}
